www/app/blog/2025-04-23-drop-column.mdx

```diff
@@ -0,0 +1,264 @@
+
+If you have access to a Postgres database, you can follow along. If not, you can spin one up locally, try Nile's [test container](https://www.thenile.dev/docs/getting-started/postgres_docker), or create a database on [Nile](https://console.thenile.dev).
+
+## A Riddle
+
+Let's start with a puzzle:
+
+> We have a table with 2 columns: `id` and `name`.  
+> What will happen if we add and remove a column 2000 times?
+
+Think about it for a moment and then you can try it yourself with a simple script:
+
+```python
+import psycopg2, textwrap, sys
+conn = psycopg2.connect("postgresql://user:password@us-west-2.db.thenile.dev:5432/dbbame") ; cur = conn.cursor()
+conn.autocommit = True
+cur.execute("DROP TABLE IF EXISTS t; CREATE TABLE t(a int, b int);")
+for i in range(1, 2001):
+    cur.execute(textwrap.dedent(f"""
+        ALTER TABLE t ADD COLUMN c{i} int;
+        ALTER TABLE t DROP COLUMN c{i};
+    """))
+cur.close(); conn.close()
+```
+
+It looks harmless. The table should end up with just the original two columns. But instead, you'll hit this:
+
+```
+psycopg2.errors.TooManyColumns: tables can have at most 1600 columns
+```
+
+But... we only have two columns? What gives?
+
+## What's happening under the hood?
+
+[Postgres has a hard limit of 1600 columns per table](https://www.postgresql.org/docs/15/limits.html).
+We hit this limit when we only had two columns in our table because when you drop a column, it doesn't actually get deleted.
+
+### Why doesn't Postgres delete the column?
+
+Postgres stores table data in files, divided into 8KB pages. Each page holds metadata and the actual row data.
+
+![Diagram of a Postgres page](./2025-04-23-drop-column/pgdata.png)
+
+Now, imagine that we have to delete a column from the table. Physically removing the data would be very inefficient - Postgres would have to scan
+the entire file(s), and in each and every row, find the column we want to delete and either replace it with a null value or re-write the entire row without it.
+This kind of full table rewrite is very expensive, and Postgres avoids it by... not deleting the column.
+
+So instead, Postgres marks the column as "dropped" in metadata. The data remains, but Postgres will ignore the dropped column during queries, inserts, and updates. Much faster.
+
+You can see this in action by checking the `pg_attribute` table after we've added and removed a column:
+
+```sql
+CREATE TABLE test2 (a integer, b integer, c integer);
+INSERT INTO test2 VALUES (1, 2, 3);
+ALTER TABLE test2 DROP COLUMN b;
+INSERT INTO test2 VALUES (1, 3);
+
+SELECT attnum, attname, attisdropped, format_type(atttypid, atttypmod) AS data_type
+FROM   pg_attribute
+WHERE  attrelid = 'test2'::regclass;
+```
+
+The output is:
+
+```text
+ attnum |           attname            | attisdropped | data_type
+--------+------------------------------+--------------+-----------
+     -6 | tableoid                     | f            | oid
```

Comment:
nit: maybe use `AND attnum > 0` in the query so people don't focus on what are those hidden columns, and things like why postgres seems to waste disk space storing the table oid for each row (spoiler: it actually doesn't).

Reply:
LOL. I thought the negative attnum are pretty nifty, but I was also wondering why store tableoid for each row :) Maybe a topic for another blog.